### DISCUSSION TITLE
Having the TOC also include HTML headings <h>

### DISCUSSION THREAD

**Discussion by guivnf** [squidfunk/mkdocs-material#7707] [Q&A] [answered]
Basically i've been having issues with the TOC as when creating headings that use HTML instead of the usual markdown annotations like in
```markdown
## Heading
```

What i use
```html
<h3 id="weapon-file"">Weapon File</h3>
```

They're excluded from the table of contents
![image](https://github.com/user-attachments/assets/bb8e3cb9-aa1f-4155-8590-47fb6d7d360e)

I've seen a few simmilar discussions on the same topic, but i haven't been able to reproduce the contrary expected result, like in:
https://github.com/squidfunk/mkdocs-material/issues/313

**Comment by squidfunk** [comment]
This is a limitation of [Python Markdown](https://github.com/Python-Markdown/markdown), which MkDocs uses (our upstream dependency) to render Markdown. You can see that this is not related to Material for MkDocs by using another theme – same result. I think this is a deliberate design decision. In any case, please take this upstream.

**Reply by guivnf** [reply]
Will do thanks.

**Comment by facelessuser** [comment]
I think there is an important misunderstanding being made. Yes, this is a limitation in Python Markdown, but you can force raw HTML headers to be included in the TOC by using `md_in_html`.

Normally, HTML is taken out and handled differently. But if you use `md_in_html` and attach the `markdown` attribute, it will be handled more like normal Markdown.

```py
import markdown
import pymdownx

print(f'Pymdown Extensions Version: {pymdownx.__version__}')
print(f'Markdown Version: {markdown.__version__}')
print('-----')

MD = """
[TOC]

## Test 1

<h2 markdown>Test 2</h2>
"""

extensions = ["toc", "md_in_html"]
extension_configs = {}

print(markdown.markdown(MD, extensions=extensions, extension_configs=extension_configs))
```

```html
<div class="toc">
<ul>
<li><a href="#test-1">Test 1</a></li>
<li><a href="#test-2">Test 2</a></li>
</ul>
</div>
<h2 id="test-1">Test 1</h2>
<h2 id="test-2">Test 2</h2>
```

Are there cases where this may not always work? Maybe, but generally, at root level, this should work.

**Reply by guivnf** [reply]
> I think there is an important misunderstanding being made. Yes, this is a limitation in Python Markdown, but you can force raw HTML headers to be included in the TOC by using `md_in_html`.

Wouldn't forcing HTML headers into the TOC require manually specifying every single anchor?

**Reply by facelessuser** [reply]
I don't know what you are asking. Maybe an example will help?

**Reply by guivnf** [reply]
> I don't know what you are asking. Maybe an example will help?

Ok so from my understanding, you said that there could be a way to force raw HTML headers to be included in the TOC, by using the ``md_in_html`` markdown extension, but for that i'd need to completely override the TOC and to manually specify the anchor links onto it.

 My question was, by doing this override, to include the anchors, wouldn't that completely bust the other markdown headers (the ones that use #) and force me to have to manually write down each anchor so it's included in the TOC? 
 
 Idk how i could provide an example of this, but yeah, correct me if i'm wrong but i don't think i fully get your solution.

**Reply by squidfunk** [reply]
@facelessuser I wasn't aware of this workaround, thanks for sharing

**Reply by guivnf** [reply]
> I don't know what you are asking. Maybe an example will help?

Ok so i already have md_in_html enabled, here's what i have
```markdown
<h3 markdown id="weapon-file">Weapon File</h3>
```
and it still doesn`t show up in the toc as you may see in the image below:
![image](https://github.com/user-attachments/assets/1e233026-48f3-4ea7-8e87-271ff4dc6a2c)

Here's what inside of my mkdocs.yml
![image](https://github.com/user-attachments/assets/7e716438-0d0f-48d0-a923-cf2257af21ab)

I'm just really lost here, because i've also tried to use the examples you showed me. And they do work, but they also look like this:
![image](https://github.com/user-attachments/assets/d5c316b9-3bf6-429d-be8b-58bdee4aab52)

**Reply by facelessuser** [reply]
> but for that i'd need to completely override the TOC and to manually specify the anchor links onto it.

I'm not sure where you are getting this from. Just include `md_in_html` and use raw HTML when you want to. That's it. Maybe you are confused because my example showed me adding `[TOC]`? You don't have to do this, I just wanted to illustrate it worked.

> My question was, by doing this override, to include the anchors, wouldn't that completely bust the other markdown headers (the ones that use #) and force me to have to manually write down each anchor so it's included in the TOC?

Again, I think your assumptions are not quite right here. Try it and see what happens 🙂.

**Reply by facelessuser** [reply]
@MetalKnight56 



> and it still doesn`t show up in the toc as you may see in the image below:

It shows up for me

<img width="152" alt="Screenshot 2024-11-17 at 10 49 51 AM" src="https://github.com/user-attachments/assets/a8e9fb3f-1ca7-4eb4-87bf-f410728117db">

If you want someone to look at this further, please provide a minimal, reproducible example.

**Reply by guivnf** [reply]
> If you want someone to look at this further, please provide a minimal, reproducible example.

Try to have it wrapped inside of these 

```markdown
<div style="display: flex; align-items: flex-start;">
<div style="flex: 1; margin-right: 20px; max-width:75%;">

<h3 markdown id="weapon-file">
<blockquote>Here's a simple example of a sword config:</blockquote>
Test Test Test

</div> <div style="flex: 1; max-width:25%;">

Test 2
<br>
Test 3
</div></div>
```

It should not work, that's the exact thing i have on my side

**Reply by facelessuser** [reply, marked as answer]
> Try to have it wrapped inside of these

Well, you are not using `md_in_html` correctly.

Let me explain how this works. Python Markdown does not process block HTML _unless_ it has a `markdown` attribute.

So you want to process an `<h3>` inside a `<div>`? You need to ensure the parent elements also allow markdown processing.

```html
<div markdown style="display: flex; align-items: flex-start;">
<div markdown style="flex: 1; margin-right: 20px; max-width:75%;">

<h3 markdown id="weapon-file">
<blockquote>Here's a simple example of a sword config:</blockquote>
Test Test Test

</div> <div style="flex: 1; max-width:25%;">

Test 2
<br>
Test 3
</div></div>
```

<img width="224" alt="Screenshot 2024-11-17 at 10 59 11 AM" src="https://github.com/user-attachments/assets/de36b9a0-7106-4e92-a924-e2c96c567c76">

**Reply by facelessuser** [reply]
Also, read the documentation for the extension: https://python-markdown.github.io/extensions/md_in_html/.

Keep in mind that it is good to be strict when you are expecting the plugin to process your content. It epects the opening tag on a line by itself and the closing on a line by itself.
